In [1]:
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

## **Data Understanding**

Dataset supermarket adalah data perusahaan ritel yang berisi data frekuensi **platform**(*web,catalog,store*) transaksi dan pendapatan per **lini produknya**(*wines, fruits, meat, fish, sweet, gold*), disertai dengan **data pribadi** konsumennya.

In [3]:
df= pd.read_excel('cleanSuperstore.xlsx')
df

ID  Year_Birth   Generation   Education Marital_Status    Income  \
0      5524        1957  Baby Boomer  Graduation         Single   58138.0   
1      2174        1954  Baby Boomer  Graduation         Single   46344.0   
2      4141        1965        Gen X  Graduation       Together   71613.0   
3      6182        1984  Millennials  Graduation       Together   26646.0   
4      5324        1981  Millennials         PhD       Together   58293.0   
...     ...         ...          ...         ...            ...       ...   
2017   9432        1977        Gen X  Graduation       Together  666666.0   
2018  10870        1967        Gen X  Graduation       Together   61223.0   
2019   7270        1981  Millennials  Graduation       Divorced   56981.0   
2020   8235        1956  Baby Boomer      Master       Together   69245.0   
2021   9405        1954  Baby Boomer         PhD       Together   52869.0   

      Kidhome  Teenhome Dt_Customer  Recency  ...  NumStorePurchases  \
0           0         0  2012-09-04       58  ...                  4   
1           1         1  2014-03-08       38  ...                  2   
2           0         0  2013-08-21       26  ...                 10   
3           1         0  2014-02-10       26  ...                  4   
4           1         0  2014-01-19       94  ...                  6   
...       ...       ...         ...      ...  ...                ...   
2017        1         0  2013-06-02       23  ...                  3   
2018        0         1  2013-06-13       46  ...                  4   
2019        0         0  2014-01-25       91  ...                 13   
2020        0         1  2014-01-24        8  ...                 10   
2021        1         1  2012-10-15       40  ...                  4   

      Frequency  NumWebVisitsMonth  AcceptedCmp1  AcceptedCmp2  AcceptedCmp3  \
0            22                  7             0             0             0   
1             4                  5             0             0             0   
2            20                  4             0             0             0   
3             6                  6             0             0             0   
4            14                  5             0             0             0   
...         ...                ...           ...           ...           ...   
2017          7                  6             0             0             0   
2018         16                  5             0             0             0   
2019         18                  6             0             0             0   
2020         21                  3             0             0             0   
2021          8                  7             0             0             0   

      AcceptedCmp4  AcceptedCmp5  Response  Complain  
0                0             0         1         0  
1                0             0         0         0  
2                0             0         0         0  
3                0             0         0         0  
4                0             0         0         0  
...            ...           ...       ...       ...  
2017             0             0         0         0  
2018             0             0         0         0  
2019             1             0         0         0  
2020             0             0         0         0  
2021             0             0         1         0  

[2022 rows x 30 columns]

In [27]:
# deep info
stats = []
for column in df.columns:
    dtype = df[column].dtype
    null_count = df[column].isnull().sum()
    null_pct = round((null_count / len(df)) * 100,2)
    unique_count = df[column].nunique()
    unique_sample = df[column].dropna().sample(min(2, unique_count)).tolist()

    stats.append({
        'Kolom Data': column,
        'Tipe Data': dtype,
        'Null': null_count,
        'nullPct(%)': null_pct,
        'unique': unique_count,
        'uniqueSample': unique_sample
    })

pd.DataFrame(stats)

Kolom Data       Tipe Data  Null  nullPct(%)  unique  \
0                    ID           int64     0         0.0    2022   
1            Year_Birth           int64     0         0.0      56   
2            Generation          object     0         0.0       4   
3             Education          object     0         0.0       4   
4        Marital_Status          object     0         0.0       3   
5                Income         float64     0         0.0    1971   
6               Kidhome           int64     0         0.0       3   
7              Teenhome           int64     0         0.0       3   
8           Dt_Customer  datetime64[ns]     0         0.0     663   
9               Recency           int64     0         0.0     100   
10             MntWines           int64     0         0.0     775   
11            MntFruits           int64     0         0.0     158   
12      MntMeatProducts           int64     0         0.0     557   
13      MntFishProducts           int64     0         0.0     182   
14     MntSweetProducts           int64     0         0.0     177   
15         MntGoldProds           int64     0         0.0     213   
16             Monetary           int64     0         0.0    1052   
17    NumDealsPurchases           int64     0         0.0      15   
18      NumWebPurchases           int64     0         0.0      15   
19  NumCatalogPurchases           int64     0         0.0      14   
20    NumStorePurchases           int64     0         0.0      14   
21            Frequency           int64     0         0.0      32   
22    NumWebVisitsMonth           int64     0         0.0      15   
23         AcceptedCmp1           int64     0         0.0       2   
24         AcceptedCmp2           int64     0         0.0       2   
25         AcceptedCmp3           int64     0         0.0       2   
26         AcceptedCmp4           int64     0         0.0       2   
27         AcceptedCmp5           int64     0         0.0       2   
28             Response           int64     0         0.0       2   
29             Complain           int64     0         0.0       2   

                                  uniqueSample  
0                                  [4149, 564]  
1                                 [1950, 1983]  
2                   [Baby Boomer, Millennials]  
3                              [Basic, Master]  
4                         [Together, Together]  
5                           [22304.0, 40246.0]  
6                                       [2, 1]  
7                                       [0, 0]  
8   [2013-01-02 00:00:00, 2012-11-26 00:00:00]  
9                                     [11, 25]  
10                                    [430, 9]  
11                                     [6, 93]  
12                                   [161, 11]  
13                                    [11, 80]  
14                                     [0, 19]  
15                                    [24, 56]  
16                                   [20, 253]  
17                                      [2, 2]  
18                                      [4, 4]  
19                                      [5, 8]  
20                                      [4, 5]  
21                                    [21, 14]  
22                                      [2, 8]  
23                                      [0, 0]  
24                                      [0, 0]  
25                                      [0, 0]  
26                                      [0, 0]  
27                                      [0, 0]  
28                                      [0, 0]  
29                                      [0, 0]

In [70]:
df.columns

Index(['ID', 'Year_Birth', 'Generation', 'Education', 'Marital_Status',
       'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'MntWines',
       'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'Monetary', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'Frequency',
       'NumWebVisitsMonth', 'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3',
       'AcceptedCmp4', 'AcceptedCmp5', 'Response', 'Complain'],
      dtype='object')

## **Data Prediction**

In [72]:
dff=df[['ID','Kidhome', 'Teenhome','MntWines',
       'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'Monetary', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'Frequency']]

Pada project ini akan melakukan 2 Prediksi linear regresi yaitu:
1. Prediksi Frekuensi Kunjungan Ke Supermarket Terhadap Jumlah Anak (`Kidhome`, `Teenhome`, `Frequency`)
2. Prediksi Nominal transaksi konsumen Supermarket Terhadap Jumlah Anak (`Kidhome`, `Teenhome`, `Monetary`)

### **Prediksi Frekuensi Kunjungan Ke Supermarket Terhadap Jumlah Anak**

In [75]:
#memprediksi kedatangan ke supermarket berdasarkan jumlah anak
pred1= dff[['Kidhome','Teenhome','Frequency']]
pred1.head()

Kidhome  Teenhome  Frequency
0        0         0         22
1        1         1          4
2        0         0         20
3        1         0          6
4        1         0         14

In [81]:
reg_model = sm.OLS(pred1['Frequency'], sm.add_constant(pred1[['Kidhome','Teenhome']]))
hasil = reg_model.fit()
print(hasil.summary())

                            OLS Regression Results                            
Dep. Variable:              Frequency   R-squared:                       0.332
Model:                            OLS   Adj. R-squared:                  0.331
Method:                 Least Squares   F-statistic:                     501.7
Date:                Mon, 18 Sep 2023   Prob (F-statistic):          1.28e-177
Time:                        15:45:18   Log-Likelihood:                -6444.1
No. Observations:                2022   AIC:                         1.289e+04
Df Residuals:                    2019   BIC:                         1.291e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         15.9759      0.211     75.607      0.0

Dari Hasil diatas terdapat beberapa point yang bisa dijabarkan:
+ **R-squared** senilai 0.332 yang berarti **33.2%** jumlah anak dan remaja mempengaruhi frekuensi kunjungan konsumen supermarket
+ **Prob (F-statistic)** senilai 1.28e-177 yang lebih kecil dari 0.05 sehingga **tolak H0** yang berarti nilai beta mempengaruhi secara signifikan terhadap model
+ **Partial Test**<br>
    -`Kidhome` Memiliki nilai p-value (0.000) yang lebih kecil dari 0.05 sehingga **tolak H0** yang berarti nilai beta mempengaruhi secara signifikan terhadap model. Ketika jumlah *kid* bertambah frequency kunjungan berkurang -7.6754.<br>
    -`Teenhome` Memiliki nilai p-value (0.823) yang lebih besar dari 0.05 sehingga **gagal tolak H0** yang berarti jumlah *teen* tidak mempengaruhi secara signifikan terhadap model.

Persamaan model regresi yang didapatkan adalah
$$Y = 15.9759 + (- 7.6754x)$$

Y = `Monetary`<br>
x = `Kidhome`<br>

In [4]:
def model(x):
    Y = 15.9759 + (-7.6754*x)
    return Y

model(2)

0.6250999999999998

### **Prediksi Total Belanja Konsumen Terhadap Jumlah Anak**

In [83]:
#jumlah anak dan remaja dalam rumah terhadap kedatangan ke supermarket
pred2= df[['Kidhome','Teenhome','Monetary']]
pred2.head()

Kidhome  Teenhome  Monetary
0        0         0      1617
1        1         1        27
2        0         0       776
3        1         0        53
4        1         0       422

In [52]:
reg_model = sm.OLS(pred2['Monetary'], sm.add_constant(pred2[['Kidhome','Teenhome']]))
hasil = reg_model.fit()
print(hasil.summary())

                            OLS Regression Results                            
Dep. Variable:               Monetary   R-squared:                       0.344
Model:                            OLS   Adj. R-squared:                  0.343
Method:                 Least Squares   F-statistic:                     528.4
Date:                Mon, 18 Sep 2023   Prob (F-statistic):          2.84e-185
Time:                        15:02:25   Log-Likelihood:                -15385.
No. Observations:                2022   AIC:                         3.078e+04
Df Residuals:                    2019   BIC:                         3.079e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        985.1794     17.586     56.022      0.0

Dari Hasil diatas terdapat beberapa point yang bisa dijabarkan:
+ **R-squared** senilai 0.344 yang berarti **34.4%** jumlah anak dan remaja mempengaruhi nominal belanja konsumen supermarket
+ **Prob (F-statistic)** senilai 2.84e-185 yang lebih kecil dari 0.05 sehingga **tolak H0** yang berarti nilai beta mempengaruhi secara signifikan terhadap model
+ **Partial Test**<br>
    -`Kidhome` Memiliki nilai p-value (0.000) yang lebih kecil dari 0.05 sehingga **tolak H0** yang berarti nilai beta mempengaruhi secara signifikan terhadap model. Ketika jumlah *kid* bertambah nominal belanja berkurang -637.8138 <br>
    -`Teenhome` Memiliki nilai p-value (0.00)yang lebih kecil dari 0.05 sehingga **tolak H0** yang berarti nilai beta mempengaruhi secara signifikan terhadap model. Ketika jumlah *teen* bertambah nominal belanja berkurang -182.0574<br>

Persamaan model regresi yang didapatkan adalah
$$Y = 985.1794 + (-637.8138x_1) + (-182.0574x_2)$$

Y = `Monetary`<br>
x1 = `Kidhome`<br>
x2 = `Teenhome`<br>

In [9]:
def model(x1,x2):
    Y = 985.1794 + (-637.8138*x1) + (-182.0574*x2)
    return Y

model(2,2)

-654.5630000000001